docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,247 @@
+-------------------------------------             |         |
+| HostStatusReclaimingPendingReboot |------       |         |
+-------------------------------------     |       |         |
+                |                         |       |         |
+  ------------------------------       Timeout    |         |
+  | TransitionTypeRegisterHost |          |       |         |
+  ------------------------------          |       |         |
+                |                         |       |         |
+                V                         V       V         V
+--------------------------------  ----------------------------------------
+| HostStatusDiscoveringUnbound |  | HostStatusUnbindingPendingUserAction |
+--------------------------------  ----------------------------------------
+              ^                                       |
+              |       ------------------------------  |
+              --------| TransitionTypeRegisterHost |---
+                      ------------------------------
+```
+
+#### Assumptions
+
+* This approach only works if the host is currently part of a working cluster
+  - This cluster needs to have been installed by assisted service so that
+    it has access to the kubeconfig
+* This reclaim procedure will only be triggered if assisted service is running
+  the kubernetes API controllers
+  - Assisted service needs to ensure the host isn't being managed by a bare
+    metal host (only possible with the kubernetes API)
+  - When using the REST API there is no way for a user to indicate if a host
+    should be unbound normally or if it should be reclaimed
+* There must be enough free space on the `/boot` partition
+  - Space is required to store the initrd and kernel (~90MB)
+  - The agent can check for this
+* A node in a spoke cluster can be identified by the hostname in the agent
+  - In a non-CAPI scenario there are no other reliable means to identify which
+    node maps to which agent
+  - With CAPI a string of references exist that could be used to identify the
+    node, but the hostname should be reliable in this case as well
+
+### Risks and Mitigations
+
+It's possible that something could go wrong during the boot process. A user
+that doesn't have access to the host infrastructure wouldn't be able to debug
+or solve this type of problem.
+
+A timeout will be implemented when the service expects the host to reregister
+(in `HostStatusReclaimingPendingReboot`). After this time the host will move to
+the existing `HostStatusUnbindingPendingUserAction` state.
+
+If the node hasn't been removed from the cluster when the host reboots the host
+could boot back into the node and the pod running the agent would attempt to
+register. The assisted service would not know the difference between this
+scenario and if the node had booted into discovery. To deal with this, the
+next-step-runner process will be started in the pod rather than the full agent
+as the only difference is the registration step.
```

Comment:
@mhrivnak I know we talked about this not being an issue yesterday, but I think we were focused on CAPI where the node would have been removed from the cluster.

If we're not using CAPI and the user hasn't removed the node from the cluster and it reboots, it would get its workloads back, right? If so, that would definitely cause a problem.

Reply:
I thought the [first step](https://github.com/openshift/assisted-service/pull/3849/files#diff-064572287176be5e9df4e402623ab95cce4ab3061a918d14093b7d6d891d1317R59) was that the user has to remove the host from the cluster, unless is there an instance where they wouldn't and the agent starts rebooting? Like if there were a timing issue where the service hasn't saved that the host was removed from the cluster, but the host has started/finished rebooting?

Reply:
"unbind" doesn't actually mean "remove node from cluster". And in-fact this can't be the case because if the node was removed from the cluster then after that point we wouldn't be able to start the pod that is going to actually do the work.

Reply:
Oh gotcha, thanks for the explanation!

Reply:
Looping back, does that mean that the user would need to unbind, wait for the hosts to move to state `HostStatusReclaimingPendingReboot`, then remove them from the cluster via another API?

Would it be a good idea to introduce another state before this state if it's not CAPI that prompts the user to use the remove host from cluster API? A state like `HostStatusReclaimingPendingUserAction`